airbyte-cdk/python/airbyte_cdk/sources/streams/http/error_handlers/http_status_error_handler.py

```diff
@@ -0,0 +1,138 @@
+# Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+
+import logging
+from typing import Mapping, Optional, Type, Union
+
+import requests
+from airbyte_cdk.models import FailureType
+from requests import RequestException
+
+from .error_handler import ErrorHandler
+from .response_models import ErrorResolution, ResponseAction
+
+
+class HttpStatusErrorHandler(ErrorHandler):
+
+    _DEFAULT_ERROR_MAPPING: Mapping[Union[int, str, Type[Exception]], ErrorResolution] = {
+        RequestException: ErrorResolution(
+            response_action=ResponseAction.RETRY,
+            failure_type=FailureType.transient_error,
+            error_message="An exception occurred when making the request.",
+        ),
+        400: ErrorResolution(
+            response_action=ResponseAction.FAIL,
+            failure_type=FailureType.config_error,
+            error_message="Bad request. Please check your request parameters.",
+        ),
+        401: ErrorResolution(
+            response_action=ResponseAction.FAIL,
+            failure_type=FailureType.config_error,
+            error_message="Unauthorized. Please ensure you are authenticated correctly.",
+        ),
+        403: ErrorResolution(
+            response_action=ResponseAction.FAIL,
+            failure_type=FailureType.config_error,
+            error_message="Forbidden. You don't have permission to access this resource.",
+        ),
+        404: ErrorResolution(
+            response_action=ResponseAction.FAIL,
+            failure_type=FailureType.system_error,
+            error_message="Not found. The requested resource was not found on the server.",
+        ),
+        408: ErrorResolution(
+            response_action=ResponseAction.RETRY,
+            failure_type=FailureType.transient_error,
+            error_message="Request timeout.",
+        ),
+        429: ErrorResolution(
+            response_action=ResponseAction.RETRY,
+            failure_type=FailureType.transient_error,
+            error_message="Too many requests.",
+        ),
+        500: ErrorResolution(
+            response_action=ResponseAction.RETRY,
+            failure_type=FailureType.transient_error,
+            error_message="Internal server error.",
+        ),
+        502: ErrorResolution(
+            response_action=ResponseAction.RETRY,
+            failure_type=FailureType.transient_error,
+            error_message="Bad gateway.",
+        ),
+        503: ErrorResolution(
+            response_action=ResponseAction.RETRY,
+            failure_type=FailureType.transient_error,
+            error_message="Service unavailable.",
+        ),
```

Comment:
Should we retry on 504 as well? This is the current behavior today

Reply:
Let me add on that: the current logic is to retry anything >= 500. I see Shopify would benefit from keeping this default logic as seen [here](https://shopify.dev/docs/api/usage/response-codes)
![image](https://github.com/airbytehq/airbyte/assets/3360483/2690ad71-f395-4247-b8e5-70a6351e400f)

I'm wondering if we should keep this logic

Reply:
Added retry for 504.